File: server/src/main/java/com/expenseapp/company/service/CompanyService.java
package com.expenseapp.company.service;

import com.expenseapp.company.domain.Company;
import com.expenseapp.company.repository.CompanyRepository;
import com.expenseapp.company.web.dto.CompanyRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
@Transactional
public class CompanyService {
    private final CompanyRepository repo;

    public CompanyService(CompanyRepository repo) {
        this.repo = repo;
    }

    public Company create(CompanyRequest req, Long userId) {
        if (repo.existsByCompanyNameIgnoreCase(req.getCompanyName())) {
            throw new IllegalArgumentException("companyName already exists");
        }
        if (repo.existsByCompanyCodeIgnoreCase(req.getCompanyCode())) {
            throw new IllegalArgumentException("companyCode already exists");
        }
        Company c = Company.builder()
                .companyName(req.getCompanyName().trim())
                .companyCode(req.getCompanyCode().trim())
                .registrationNumber(req.getRegistrationNumber())
                .taxId(req.getTaxId())
                .industryType(req.getIndustryType())
                .companyEmail(req.getCompanyEmail().toLowerCase())
                .contactNumber(req.getContactNumber())
                .website(req.getWebsite())
                .addressLine1(req.getAddressLine1())
                .addressLine2(req.getAddressLine2())
                .city(req.getCity())
                .state(req.getState())
                .country(req.getCountry())
                .postalCode(req.getPostalCode())
                .currency(req.getCurrency())
                .fiscalYearStart(req.getFiscalYearStart())
                .timeZone(req.getTimeZone())
                .companyLogoUrl(req.getCompanyLogoUrl())
                .status(req.getStatus() == null ? "ACTIVE" : req.getStatus())
                .createdBy(userId)
                .build();
        return repo.save(c);
    }

    @Transactional(readOnly = true)
    public List<Company> findAll() {
        return repo.findAll();
    }
    
    @Transactional(readOnly = true)
    public List<Company> findAllByUser(Long userId) {
        if (userId == null) {
            return List.of();
        }
        return repo.findByCreatedByAndStatusNot(userId, "DELETED");
    }

    @Transactional(readOnly = true)
    public Optional<Company> findOne(Long id) {
        return repo.findByIdAndStatusNot(id, "DELETED");
    }
    
    @Transactional(readOnly = true)
    public Optional<Company> findOneByUser(Long id, Long userId) {
        if (userId == null) {
            return Optional.empty();
        }
        return repo.findByIdAndCreatedByAndStatusNot(id, userId, "DELETED");
    }

    public Company update(Long id, CompanyRequest req) {
        Company c = findOne(id).orElseThrow(() -> new IllegalArgumentException("Company not found"));
        // Only update mutable fields
        c.setCompanyName(req.getCompanyName());
        c.setCompanyCode(req.getCompanyCode());
        c.setRegistrationNumber(req.getRegistrationNumber());
        c.setTaxId(req.getTaxId());
        c.setIndustryType(req.getIndustryType());
        c.setCompanyEmail(req.getCompanyEmail());
        c.setContactNumber(req.getContactNumber());
        c.setWebsite(req.getWebsite());
        c.setAddressLine1(req.getAddressLine1());
        c.setAddressLine2(req.getAddressLine2());
        c.setCity(req.getCity());
        c.setState(req.getState());
        c.setCountry(req.getCountry());
        c.setPostalCode(req.getPostalCode());
        c.setCurrency(req.getCurrency());
        c.setFiscalYearStart(req.getFiscalYearStart());
        c.setTimeZone(req.getTimeZone());
        c.setCompanyLogoUrl(req.getCompanyLogoUrl());
        c.setStatus(req.getStatus() == null ? c.getStatus() : req.getStatus());
        return repo.save(c);
    }

    public void deactivate(Long id) {
        Company c = findOne(id).orElseThrow(() -> new IllegalArgumentException("Company not found"));
        c.setStatus("INACTIVE");
        repo.save(c);
    }
}

File: mobile/src/api/companyService.ts
import { api, API_BASE_URL } from './client';

export interface CompanyPayload {
  companyName: string;
  companyCode: string;
  registrationNumber?: string;
  taxId?: string;
  industryType: string;
  companyEmail: string;
  contactNumber: string;
  website?: string;
  addressLine1: string;
  addressLine2?: string;
  city: string;
  state: string;
  country: string;
  postalCode: string;
  currency: string;
  fiscalYearStart?: string;
  timeZone: string;
  companyLogoUrl?: string;
  status?: 'ACTIVE' | 'INACTIVE';
}

export interface Company extends CompanyPayload {
  id: number;
  createdBy?: number;
  createdAt?: string;
  updatedAt?: string;
}

export const companiesService = {
  // Use main backend API (18080) - unified API
  async create(payload: CompanyPayload) {
    const { data } = await api.post<Company>('/api/v1/companies', payload);
    return data;
  },
  async list() {
    try {
      // Use the unified API endpoint that returns user's companies
      const { data } = await api.get<Company[]>('/api/v1/companies');
      if (Array.isArray(data)) {
        console.log('[companiesService.list] loaded companies:', data.length);
        return data;
      }
      console.log('[companiesService.list] no companies found');
      return [];
    } catch (err: any) {
      console.error('[companiesService.list] error:', err.message);
      return [];
    }
  },
  async get(id: number) {
    const { data } = await api.get<Company>(`/api/v1/companies/${id}`);
    return data;
  },
  async update(id: number, payload: Partial<CompanyPayload>) {
    const { data } = await api.put<Company>(`/api/v1/companies/${id}`, payload);
    return data;
  },
  async remove(id: number) {
    await api.delete(`/api/v1/companies/${id}`);
  },
};
